### PR TITLE
fix userschema issues

### PR DESCRIPTION
**Pull request by adizaimi**
* fix segfault when authentication is not on--return empty table
* check and return error when user and password exceed proper length
* add testcases

### CODE REVIEW

File: db/sqlinterfaces.c [hunk 1]
```diff
@@ -7755,16 +7755,36 @@ static int process_set_commands(struct sqlclntstate *clnt)
...
                 }
             } else if (strncasecmp(sqlstr, "user", 4) == 0) {
+                char *savptr = sqlstr;
                 sqlstr += 4;
                 sqlstr = cdb2_skipws(sqlstr);
-                clnt->have_user = 1;
-                strncpy(clnt->user, sqlstr, sizeof(clnt->user));
+                if (strlen(sqlstr) >= sizeof(clnt->user)) {
+                    sqlstr = savptr;
+                    logmsg(LOGMSG_USER,
```
Comment (nirbhayc) [inline db/sqlinterfaces.c]: I’d expect a test that fails with this error.

File: tests/userschema.test/t09.expected [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+(t=2)
+(tablename='forecast@dcba', username='abcd', READ='N', WRITE='N', DDL='N')
+(tablename='forecast@dcba', username='dcba', READ='Y', WRITE='Y', DDL='Y')
+(tablename='forecast@dcba', username='default', READ='N', WRITE='N', DDL='N')
+(tablename='forecast@dcba', username='mohit', READ='Y', WRITE='Y', DDL='Y')
+(tablename='test@dcba', username='abcd', READ='N', WRITE='N', DDL='N')
+(tablename='test@dcba', username='dcba', READ='Y', WRITE='Y', DDL='Y')
+(tablename='test@dcba', username='default', READ='N', WRITE='N', DDL='N')
+(tablename='test@dcba', username='mohit', READ='Y', WRITE='Y', DDL='Y')
+[select * from test] failed with rc -3 Invalid set command 'user 12345678901234567'
```
Comment (nirbhayc) [inline tests/userschema.test/t09.expected]: Why do we get this error on SELECT?

File: db/sqlinterfaces.c [hunk 3]
```diff
@@ -7755,16 +7755,36 @@ static int process_set_commands(struct sqlclntstate *clnt)
...
+                           "set password parameter can not exceed length of %d "
+                           "characters\n", sizeof(clnt->password) - 1);
+                    rc = ii + 1;
+                } else {
+                    clnt->have_password = 1;
+                    strncpy(clnt->password, sqlstr, sizeof(clnt->password) - 1);
+                    clnt->password[sizeof(clnt->password) - 1] = '\0';
+                    sqlite3Dequote(clnt->password);
```
Comment (nirbhayc) [inline db/sqlinterfaces.c]: I think the password should be dequoted first before validating its length. Ditto for username.

File: tests/userschema.test/t09.req [hunk 4]
```diff
@@ -0,0 +1,26 @@
+select * from test
+select * from comdb2_tablepermissions order by tablename
+
+set user dcba
```
Comment (nirbhayc) [inline tests/userschema.test/t09.req]: Can you also add test for quoted user/password as well?

File: tests/userschema.test/t09.expected (hunk 2 above)
Comment (adizaimi) [reply tests/userschema.test/t09.expected]: We don't send anything to server until the select.

File: tests/userschema.test/t09.req (hunk 4 above)
Comment (adizaimi) [reply tests/userschema.test/t09.req]: ok.

File: db/sqlinterfaces.c (hunk 3 above)
Comment (adizaimi) [reply db/sqlinterfaces.c]: good point.

File: db/sqlinterfaces.c (hunk 1 above)
Comment (adizaimi) [reply db/sqlinterfaces.c]: This is printed on the .db log file. It would be nice to return that to user.

File: db/sqlinterfaces.c (hunk 3 above)
Comment (adizaimi) [reply db/sqlinterfaces.c]: Good idea. Will apply.

File: db/sqlinterfaces.c (hunk 1 above)
Comment (adizaimi) [reply db/sqlinterfaces.c]: Changing errstr to return that.

File: db/sqlinterfaces.c [hunk 5]
```diff
@@ -7755,16 +7757,35 @@ static int process_set_commands(struct sqlclntstate *clnt)
...
-                strncpy(clnt->user, sqlstr, sizeof(clnt->user));
+                sqlite3Dequote(sqlstr);
+                if (strlen(sqlstr) >= sizeof(clnt->user)) {
+                    snprintf(err, sizeof(err), "set user: '%s' exceeds %d characters",
+                           sqlstr, sizeof(clnt->user) - 1);
+                    sqlstr = savptr;
+                    rc = ii + 1;
+                } else {
+                    clnt->have_user = 1;
+                    strncpy(clnt->user, sqlstr, sizeof(clnt->user) - 1);
```
Comment (nirbhayc) [inline db/sqlinterfaces.c]: strcpy should be safe here, we have already checked the size. That makes the below line redundant too. Ditto for password.

File: sqlite/ext/comdb2/tablepermissions.c [hunk 6]
```diff
@@ -98,6 +98,8 @@ static int permissionsOpen(sqlite3_vtab *p, sqlite3_vtab_cursor **ppCursor){
   bdb_user_get_all(&pCur->ppUsers, &pCur->nUsers);
   unlock_schema_lk();
   *ppCursor = &pCur->base;
+  if (pCur->nUsers == 0)
+      return SQLITE_EMPTY;
```
Comment (nirbhayc) [inline sqlite/ext/comdb2/tablepermissions.c]: This isn’t right. IMHO, an empty table is perfectly normal, and shouldn’t be treated as error.

File: tests/userschema.test/t09.expected (hunk 2 above)
Comment (nirbhayc) [reply tests/userschema.test/t09.expected]: Huh..? On failure, I’d expect the SET command to report an error. But, it should probably be dealt in a separate PR.

File: sqlite/ext/comdb2/tablepermissions.c (hunk 6 above)
Comment (adizaimi) [reply sqlite/ext/comdb2/tablepermissions.c]: Well, this table can only have content when user schema is turned on. This was meant to be like an indicator for that. But I agree that this looks not too good.

**Comment by adizaimi** [comment]
Thank you Nirbhay. Appreciate your input.

**Pull request merged**